src/Native/FactorizationMachineNative/FactorizationMachineCore.cpp

```diff
@@ -0,0 +1,168 @@
+#include <limits>
+#include <pmmintrin.h>
+#include "../Stdafx.h"
+
+// Compute the output value of the field-aware factorization, as the sum of the linear part and the latent part. 
+// The linear part is the inner product of linearWeights and featureValues.
+// The latent part is the sum of all intra-field interactions in one field f, for all fields possible. 
+EXPORT_API(void) CalculateIntermediateVariablesNative(int fieldCount, int latentDim, int count, _In_ int * fieldIndices, _In_ int * featureIndices, _In_ float * featureValues,
+    _In_ float * linearWeights, _In_ float * latentWeights, _Inout_ float * latentSum, _Out_ float * response)
+{
+	// The number of all possible fields
+    const int m = fieldCount;
+    const int d = latentDim;
+    const int c = count;
+    const int * pf = fieldIndices;
+    const int * pi = featureIndices;
+    const float * px = featureValues;
+    const float * pw = linearWeights;
+    const float * pv = latentWeights;
+    float * pq = latentSum;
+    float linearResponse = 0;
+    float latentResponse = 0;
+
+    memset(pq, 0, sizeof(float) * m * m * d);
+    __m128 _y = _mm_setzero_ps();
+    __m128 _tmp = _mm_setzero_ps();
+
+    for (int i = 0; i < c; i++)
+    {
+        const int f = pf[i];
+        const int j = pi[i];
+        linearResponse += pw[j] * px[i];
+
+        const __m128 _x = _mm_load1_ps(px + i);
+        const __m128 _xx = _mm_mul_ps(_x, _x);
+
+        // tmp -= <v_j,f, v_j,f> * x * x
+        const int vBias = j * m * d + f * d;
+
+		// j-th feature's latent vector in the f-th field hidden space.
+        const float * vjf = pv + vBias;
+
+        for (int k = 0; k + 4 <= d; k += 4)
+        {
+            const __m128 _v = _mm_load_ps(vjf + k);
+            _tmp = _mm_sub_ps(_tmp, _mm_mul_ps(_mm_mul_ps(_v, _v), _xx));
+        }
+
+        for (int fprime = 0; fprime < m; fprime++)
+        {
+            const int vBias = j * m * d + fprime * d;
+            const int qBias = f * m * d + fprime * d;
+            const float * vjfprime = pv + vBias;
+            float * qffprime = pq + qBias;
+
+            // q_f,f' += v_j,f' * x
+            for (int k = 0; k + 4 <= d; k += 4)
+            {
+                const __m128 _v = _mm_load_ps(vjfprime + k);
+                __m128 _q = _mm_load_ps(qffprime + k);
+                _q = _mm_add_ps(_q, _mm_mul_ps(_v, _x));
+                _mm_store_ps(qffprime + k, _q);
+            }
+        }
+    }
+
+    for (int f = 0; f < m; f++)
+    {
+        // tmp += <q_f,f, q_f,f>
+        const float * qff = pq + f * m * d + f * d;
+        for (int k = 0; k + 4 <= d; k += 4)
+        {
+            __m128 _qff = _mm_load_ps(qff + k);
+            
+			// Intra-field interactions. 
+            _tmp = _mm_add_ps(_tmp, _mm_mul_ps(_qff, _qff));
+        }
+
+        // y += <q_f,f', q_f',f>, f != f'
+        // We only handle inter - field interactions because f != f'.
+        for (int fprime = f + 1; fprime < m; fprime++)
```

Comment:
Sounds like intra-field interactions are ignored. May be change it to "This loop handles ..."

Reply:
1. Whis-->This
2. There should be no space between inter, -, and field.